wcfsetup/install/files/lib/system/package/plugin/FileDeletePackageInstallationPlugin.class.php

```diff
@@ -0,0 +1,274 @@
+                ->nullable(),
+        ]);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function fetchElementData(\DOMElement $element, $saveData)
+    {
+        return [
+            'application' => $element->getAttribute('application') ?? 'wcf',
+            'file' => $element->nodeValue,
+            'packageID' => $this->installation->getPackage()->packageID,
+        ];
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getElementIdentifier(\DOMElement $element)
+    {
+        $app = $element->getAttribute('application') ?? 'wcf';
+
+        return \sha1($app . '_' . $element->nodeValue);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function setEntryListKeys(IDevtoolsPipEntryList $entryList)
+    {
+        $entryList->setKeys([
+            'file' => 'wcf.acp.pip.fileDelete.file',
+            'application' => 'wcf.acp.pip.fileDelete.application',
+        ]);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function insertNewXmlElement(XML $xml, \DOMElement $newElement)
+    {
+        $delete = $xml->xpath()->query('/ns:data/ns:delete')->item(0);
+        if ($delete === null) {
+            $data = $xml->xpath()->query('/ns:data')->item(0);
+            $delete = $xml->getDocument()->createElement('delete');
+            DOMUtil::prepend($delete, $data);
+        }
+
+        $delete->appendChild($newElement);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function prepareXmlElement(\DOMDocument $document, IFormDocument $form)
+    {
+        $file = $document->createElement($this->tagName);
+
+        $data = $form->getData()['data'];
+        if (!empty($data['application'])) {
+            $file->setAttribute('application', $data['application']);
+        }
+        $file->nodeValue = $data['file'];
+
+        return $file;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function prepareDeleteXmlElement(\DOMElement $element)
+    {
+        return null;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function saveObject(\DOMElement $newElement, ?\DOMElement $oldElement = null)
+    {
+        $newElementData = $this->getElementData($newElement, true);
+
+        $this->handleDelete([[
+            'attributes' => [
+                'application' => $newElementData['application'],
+            ],
+            'value' => $newElementData['file'],
+        ]]);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function deleteObject(\DOMElement $element)
+    {
+        // Reverting file deletions is not supported. Use the `file` PIP instead.
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function getImportElements(\DOMXPath $xpath)
+    {
+        return $xpath->query('/ns:data/ns:delete/ns:' . $this->tagName);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    protected function getEmptyXml()
+    {
+        $xsdFilename = $this->getXsdFilename();
+        $apiVersion = WSC_API_VERSION;
```

Comment:
Is this correct? Probably affects all PIPs.

Reply:
I moved this into a separate issue https://github.com/WoltLab/WCF/issues/4268.